src/hotspot/share/gc/shenandoah/shenandoahEvacTracker.cpp

```diff
@@ -23,27 +23,31 @@
  */
 #include "precompiled.hpp"
 
+#include "gc/shenandoah/shenandoahAgeCensus.hpp"
 #include "gc/shenandoah/shenandoahHeap.inline.hpp"
 #include "gc/shenandoah/shenandoahEvacTracker.hpp"
 #include "gc/shenandoah/shenandoahThreadLocalData.hpp"
 #include "gc/shenandoah/shenandoahRootProcessor.hpp"
 #include "runtime/threadSMR.inline.hpp"
 #include "runtime/thread.hpp"
 
-ShenandoahEvacuationStats::ShenandoahEvacuationStats()
+ShenandoahEvacuationStats::ShenandoahEvacuationStats(bool generational)
   : _evacuations_completed(0), _bytes_completed(0),
     _evacuations_attempted(0), _bytes_attempted(0),
-    _age_table(false) {}
+    _generational(generational),
+    _age_table(false) {
+}
 
 void ShenandoahEvacuationStats::begin_evacuation(size_t bytes) {
   ++_evacuations_attempted;
   _bytes_attempted += bytes;
 }
 
-void ShenandoahEvacuationStats::end_evacuation(size_t bytes, uint age) {
+void ShenandoahEvacuationStats::end_evacuation(size_t bytes, bool young_gen, uint age) {
   ++_evacuations_completed;
   _bytes_completed += bytes;
-  if (age > 0) {
+  if (_generational && GenShenCensusAtEvac && young_gen) {
+    assert(age > 0, "Error");
```

Comment:
Check that age is recorded _after_ incrementing at evacuation.

Reply:
Yes, it is on the path where we evacuate. It isn't for the case where it's on the marking path. (This is of course the evacuation path here.)

Reply:
It turns out that the assertion `(age > 0)` is still too strong because it might not be an aging cycle, even though we might be evacuating age 0 objects (without, in this case, incrementing their age).